conman/routes/models.py

```diff
@@ -74,6 +84,64 @@ def handle(self, request, path):
         # Deal with the request
         return handler.handle(request, path)
 
+    def move_to(self, new_url, *, move_children):
+        """
+        Move this Route to a new url.
+
+        Requires `move_children` as a keyword argument. If True, all child
+        Routes will be moved along with this parent. Otherwise, when False, any
+        and all children will remain where they are.
+
+        In the event of a clash, an IntegrityError will be raised.
+        """
+        if move_children:
+            Route.objects.move_branch(self.url, new_url)
+            self.url = new_url
+        else:
+            old_url, self.url = self.url, new_url
+            try:
+                self.save()
+            except IntegrityError:
+                self.url = old_url
+                raise
+
+    def swap_with(self, other_route, *, move_children):
+        """
+        Swap this Route's URL with that of another Route.
+
+        Requires `move_children` as a keyword argument. When `True`, all
+        descendant routes will be moved along with the parents. When `False`,
+        they will remain where they are, and only the parents will be swapped.
+
+        When one of the `Routes` is a descendant of the other, `move_children`
+        must be `False`.
+
+        If DEFERRED unique contrstraints get supported in django, we might
+        benefit from revisiting how this works (ie: dropping the UUID).
+
+        See https://code.djangoproject.com/ticket/20581.
+        """
+        if not (self.pk and other_route.pk):
+            raise ValueError('Cannot move unsaved Routes.')
```

Comment:
I suspect we're talking about two situations here.

1. Model has been instantiated, but never saved to the database. This is what I intended to protect against. In this case, if someone has explicitly set `.pk` on the model... well... they're daft, otherwise, I think checking `pk` should be fine.

2. Model has been saved at some point, but `.url` doesn't match what's in the database. I think this is where `django-dirtyfields` might come in handy, but that seems like a fair bit of overhead to add to protect against this behaviour. I think (...hope) that it's unlikely that someone would edit the `url` manually, fail to call `save`, and the pass it into `swap_with`.